src/OpenSpaceToolkit/Physics/Environment/Atmospheric/Earth/NRLMSISE00.cpp

```diff
@@ -77,100 +85,148 @@ bool NRLMSISE00::isDefined() const
 
 Unique<NRLMSISE00::ap_array> NRLMSISE00::computeApArray(const Instant& anInstant) const
 {
-    // Algorithm:
-    //
-    // Ap readings are given on a daily basis.
-    // We first create a timestamp in each day for which we will need data.
-    // Sometimes "now" and "now-9hr" land on the same day.
-    //
-    //  now-57hr                 now
-    //     v                      v
-    //     *        *        *    *
-    //     |--------|--------|----|
-    //        24hr     24hr   9hr
-    //
-    // Then, we fetch the Ap data for each of those days
-    // For each day, we get 8 3-hourly readings, starting at 00:00 UTC.
-    // They usually do not line up with the 24-hour periods.
-    //
-    //     *        *        *    *
-    //     |--------|--------|----|
-    //|........|........|........|........|
-    //   day1     day2     day3     day4
-    //
-    // We stack the data into a single array then find which data point
-    // on day 1 lines up with the "now-57hr" timestamp
-    //
-    //     *        *        *    *
-    //     |--------|--------|----|
-    //|........|........|........|........|
-    //     ^
-    //  this one
-    //
-    // Then use that as a starting index to find the correct data points and averages.
-
-    const Manager& spaceWeatherManager = Manager::Get();
-
-    // Fetch AP parameters for each day up to 57 hours ago
-    const Instant instant57HrPrevious = anInstant - Duration::Hours(57);
-    const Instant instant33HrPrevious = anInstant - Duration::Hours(33);
-    const Instant instant9HrPrevious = anInstant - Duration::Hours(9);
-
-    // Fetch daily AP array for each day and stack end-to-end
-    Array<Instant> fetchDays = {instant57HrPrevious, instant33HrPrevious, instant9HrPrevious};
-
-    // add today if not already covered by instant9HrPrevious
-    if (instant9HrPrevious.getDateTime(Scale::UTC).getDate().getDay() !=
-        anInstant.getDateTime(Scale::UTC).getDate().getDay())
-    {
-        fetchDays.add(anInstant);
-    }
-
-    // Fetch AP indices for all days. Stack them into one continuous array.
-    Array<Integer> apMultiDayArray = Array<Integer>::Empty();
-    apMultiDayArray.reserve(fetchDays.getSize());
+    Unique<NRLMSISE00::ap_array> outputStruct = std::make_unique<NRLMSISE00::ap_array>();
 
-    for (const Instant& fetchDay : fetchDays)
+    switch (this->inputDataSourceType_)
     {
-        apMultiDayArray.add(spaceWeatherManager.getAp3HourSolarIndicesAt(fetchDay));
+        case InputDataSourceType::ConstantFluxAndGeoMag:
```

Comment:
I think this logic can be simplified a bit. The AP struct is actually allowed to be a null pointer if we're not using it. (unless a particular switch is set to -1, which we don't do)
https://github.com/magnific0/nrlmsise-00/blob/master/nrlmsise-00.h#L50

So you can probably skip the entire `computeApArray` function for the case of constant input.

Reply:
I can't comment on it directly since it's not changed, but we'll need to set this line in the input struct to point to the value of our constant Ap:
> input->ap_a = apValues.get();

Reply:
I tried the following
```
Unique<NRLMSISE00::nrlmsise_input> NRLMSISE00::computeNRLMSISE00Input(
    const Unique<NRLMSISE00::ap_array>& apValues, const LLA& aLLA, const Instant& anInstant
) const
{
    Real f107Previous = Real::Undefined();
    Real f107Average = Real::Undefined();

    switch (this->inputDataType_)
    {
        case InputDataType::ConstantFluxAndGeoMag:
        {
            // Use constant values for F10.7, F10.7a and Kp NRLMSISE00 input parameters

            f107Previous = this->f107ConstantValue_;
            f107Average = this->f107AConstantValue_;

            break;
        }

        case InputDataType::CSSISpaceWeatherFile:
        {
            // Use historical and predicted values for F10.7, F10.7a and Kp NRLMSISE00 input parameters

            // Input reference is in the NRLMSISE header file
            // https://github.com/magnific0/nrlmsise-00/blob/master/nrlmsise-00.h

            const Manager& spaceWeatherManager = Manager::Get();

            // Solar flux values
            f107Previous = spaceWeatherManager.getF107SolarFluxAt(anInstant - Duration::Days(1));
            f107Average = spaceWeatherManager.getF107SolarFlux81DayAvgAt(anInstant);

            break;
        }
    }

    const DateTime currentDateTime = anInstant.getDateTime(Scale::UTC);

    // current year/doy/sec
    const Integer year = currentDateTime.getDate().getYear();

    const Instant startOfYear = Instant::DateTime(DateTime(Date(year, 1, 1), Time::Midnight()), Scale::UTC);
    const Integer dayOfYear = (anInstant - startOfYear).getDays() + 1;

    const Time timeOfDay = currentDateTime.getTime();
    const Integer secondsInDay = timeOfDay.getHour() * 3600 + timeOfDay.getMinute() * 60 + timeOfDay.getSecond();

    Real lst = Real::Undefined();

    // Use actual sun position to compute local solar time if provided
    if (sunCelestialSPtr_)
    {
        const Position position = {
            aLLA.toCartesian(earthRadius_, earthFlattening_), Position::Unit::Meter, earthFrameSPtr_
        };

        const Position sunPositionInFrame = sunCelestialSPtr_->getPositionIn(earthFrameSPtr_, anInstant);

        lst = (Real::Pi() + std::atan2(
                                sunPositionInFrame.accessCoordinates()[0] * position.accessCoordinates()[1] -
                                    sunPositionInFrame.accessCoordinates()[1] * position.accessCoordinates()[0],
                                sunPositionInFrame.accessCoordinates()[0] * position.accessCoordinates()[0] +
                                    sunPositionInFrame.accessCoordinates()[1] * position.accessCoordinates()[1]
                            )) *
              12.0 / Real::Pi();
    }
    else
    {
        // This is the preferred method per the NRLMSISE documentation
        // https://github.com/magnific0/nrlmsise-00/blob/master/nrlmsise-00.h#L103
        lst = Real::Integer(secondsInDay) / 3600.0 + aLLA.getLongitude().inDegrees() / 15.0;
    }

    Unique<NRLMSISE00::nrlmsise_input> input = std::make_unique<NRLMSISE00::nrlmsise_input>();

    input->doy = dayOfYear;
    input->year = year;
    input->sec = secondsInDay;
    input->alt = aLLA.getAltitude().inKilometers();
    input->g_lat = aLLA.getLatitude().inDegrees();
    input->g_long = aLLA.getLongitude().inDegrees();
    input->lst = lst;
    input->f107A = f107Average;
    input->f107 = f107Previous;

    if (apValues != nullptr)
    {
        input->ap = apValues->a[0];
        input->ap_a = apValues.get();
    }
    else
    {
        const Real ap = this->convertKpToAp(this->kpConstantValue_);

        input->ap = ap;
        input->ap_a = nullptr;
    }

    return input;
}

Real NRLMSISE00::convertKpToAp(const Real& aKp)
{
    if (!aKp.isDefined())
    {
        throw ostk::core::error::runtime::Undefined("Kp");
    }

    return exp((aKp + 1.6) / 1.75);  // Exponential approximation, Vallado, 2nd edition (eq 8-31).
}

NRLMSISE00::InputDataType NRLMSISE00::getInputDataType() const
{
    return this->inputDataType_;
}

Real NRLMSISE00::getF107ConstantValue() const
{
    return this->f107ConstantValue_;
}

Real NRLMSISE00::getF107AConstantValue() const
{
    return this->f107AConstantValue_;
}

Real NRLMSISE00::getKpConstantValue() const
{
    return this->kpConstantValue_;
}

Real NRLMSISE00::getDensityAt(const LLA& aLLA, const Instant& anInstant) const
{
    Unique<NRLMSISE00::ap_array> apValues = nullptr;

    if (this->inputDataType_ == InputDataType::CSSISpaceWeatherFile)
    {
        apValues = this->computeApArray(anInstant);
    }

    const Unique<NRLMSISE00::nrlmsise_input> input = this->computeNRLMSISE00Input(apValues, aLLA, anInstant);

    return NRLMSISE00::GetDensityAt(*input);
}
```

I must be doing something wrong 🤔

Reply:
lmk